app/services/foreman/provision/ssh.rb

```diff
@@ -50,64 +80,99 @@ def command_prefix
     (username == "root") ? "" : "sudo "
   end
 
-  def command
-    # Use the users home to store the provision script since we can't reliably
-    # tell if other locations are writeable or executable by the user.
+  # Use the users home to store the provision script since we can't reliably
+  # tell if other locations are writeable or executable by the user.
+  def main_command
     main_execution = "(chmod 0701 ./#{remote_script} && #{command_prefix} ./#{remote_script} ; echo $? >#{remote_script}.status) 2>&1"
-    "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; exit $(cat #{remote_script}.status)'"
-  end
-
-  def defaults
-    {
-      :keys_only    => true,
-      :config       => false,
-      :auth_methods => %w(publickey),
-      :compression  => true,
-      :logger       => logger,
-    }
+    cleanup = "#{command_prefix} rm -f #{remote_script}"
+    "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; #{cleanup}; exit $(cat #{remote_script}.status)'"
```

Comment:
If it's executed this way you can even consider not writing the file and simply piping stdin straight to `sh` to avoid `scp` usage. Not sure how well the rest of the script likes that.

Reply:
Since I never know what can be the content of the template piped into the command, I would rather stay with the file upload.